### PR TITLE
Lvnod 105 create controller for registration

### PR DESCRIPTION
**Pull request by yurii-kokhalevych**
## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/browse/LVNOD-105)

## Code reviewers

- [x] @yarvanya
- [x] @denisbodnar
- [x] @yuliannakorol

### Second Level Review

- [ ] @lhalam
- [ ] @4asof4ik

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo
## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by denisbodnar — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

File: server/controllers/auth.controller.js [hunk 1]
```diff
@@ -18,5 +18,17 @@ module.exports = {
       }
     })
     .catch(error => res.status(401).send(error));
+  },
+  signup(req, res) {
```
Comment (4asof4ik) [inline server/controllers/auth.controller.js]: this should be in user.controller as create action

File: server/routes/index.js [hunk 2]
```diff
@@ -14,4 +14,5 @@ module.exports = app => {
   app.delete('/api/user/:id', userController.destroy);
 
   app.post('/api/auth/login', authController.login);
+  app.post('/api/auth/Signup', authController.signup);
```
Comment (4asof4ik) [inline server/routes/index.js]: create user action

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: server/controllers/user.controller.js [hunk 3]
```diff
@@ -3,12 +3,19 @@
 const User = require('../../config/db').User;
 
 module.exports = {
-  create(req, res) {
-    let assignUser = Object.assign({}, req.body);
+  signup(req, res) {
```
Comment (4asof4ik) [inline server/controllers/user.controller.js]: should be create

File: server/controllers/user.controller.js [hunk 4]
```diff
@@ -3,12 +3,19 @@
...
+    User.findOne({where: {email: req.body.email}}).then(user => {
+      if (user) {
+        res.status(422).send('That email address is already in use.');
```
Comment (4asof4ik) [inline server/controllers/user.controller.js]: into messages

File: server/controllers/user.controller.js [hunk 5]
```diff
@@ -3,12 +3,19 @@
 const User = require('../../config/db').User;
 
 module.exports = {
-  create(req, res) {
-    let assignUser = Object.assign({}, req.body);
+  signup(req, res) {
+    User.findOne({where: {email: req.body.email}}).then(user => {
```
Comment (4asof4ik) [inline server/controllers/user.controller.js]: User.findOne({
    where: {
        email: req.body.email
     }
}).then(user => {

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

File: server/routes/index.js [hunk 6]
```diff
@@ -10,7 +10,7 @@ module.exports = app => {
   app.get('/api/profile/:id', profileController.retrieve);
   app.put('/api/profile/:id', profileController.update);
 
-  app.post('/api/users', userController.create);
+  app.post('/api/users/', userController.create);
```
Comment (4asof4ik) [inline server/routes/index.js]: drop last /

File: server/controllers/user.controller.js [hunk 7]
```diff
@@ -2,24 +2,58 @@
...
+          let data = {
+            subject: message.activation,
+            img: 'activ.jpg',
+            host: req.headers.host,
+            route: constant.ROUTE.ACTIVATION,
+            email: req.body.email,
+            token: token
+          };
+          mailer(data, 'activation');
+          res.status(201).send(user);
```
Comment (4asof4ik) [inline server/controllers/user.controller.js]: why you send user?

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

File: server/controllers/user.controller.js [hunk 8]
```diff
@@ -2,24 +2,58 @@
...
 module.exports = {
   create(req, res) {
-    let assignUser = Object.assign({}, req.body);
+    User.findOne({
+      where: {
+        email: req.body.email
+      }
+    })
+    .then(user => {
+      if (user) {
```
Comment (lhalam) [inline server/controllers/user.controller.js]: update logic of a relatively new field in the User model

**Review by lhalam — COMMENTED** [review]
update logic of a relatively new field in the User model

**Review by yarvanya — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

File: server/controllers/user.controller.js [hunk 9]
```diff
@@ -2,24 +2,64 @@
...
+const constant = require('./../helper/constant');
 
 module.exports = {
   create(req, res) {
-    let assignUser = Object.assign({}, req.body);
-
-    User.create(assignUser)
-    .then(user => res.status(201).send(user))
-    .catch(error => res.status(400).send(error));
+    User.findOne({
```
Comment (4asof4ik) [inline server/controllers/user.controller.js]: the user is not created yet. what do you want to find?

File: server/controllers/user.controller.js [hunk 10]
```diff
@@ -2,24 +2,64 @@
...
+      where: {
+        email: req.body.email
+      }
+    })
+    .then(user => {
```
Comment (4asof4ik) [inline server/controllers/user.controller.js:47]: the user is undefined here

File: server/controllers/user.controller.js [hunk 11]
```diff
@@ -2,24 +2,64 @@
...
+      let assignUser = Object.assign({}, req.body);
+      let dataActivation = user => {
+        let token = jwt.sign({
+          id: user.id,
```
Comment (4asof4ik) [inline server/controllers/user.controller.js]: here undefined.id will be error

File: server/controllers/user.controller.js [hunk 12]
```diff
@@ -2,24 +2,64 @@
...
+      if (user) {
+        if (!user.is_invated) {
+          res.status(422).send(message.emailUsed);
+        } else {
+          User.updateAttributes(assignUser)
+          .then(dataActivation)
+          .catch(error => res.status(400).send(error));
+        }
+      } else {
+        User.create(assignUser)
```
Comment (4asof4ik) [inline server/controllers/user.controller.js]: user creating here

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

File: server/helper/messages.js [hunk 13]
```diff
@@ -1,7 +1,11 @@
 module.exports = {
-  'linkNotValid': 'Link are not valid or time expired!',
-  'userNotFound': 'This user is not found!',
-  'linkAlreadyActivated': 'Link has already actived!',
-  'congratulation': 'Congratulation!',
-  'eventNotFound': 'Event has not found! Try again!'
+  linkNotValid: 'Link are not valid or time expired!',
+  userNotFound: 'This user is not found!',
+  linkAlreadyActivated: 'Link has already actived!',
```
Comment (denisbodnar) [inline server/helper/messages.js]: activated

**Review by denisbodnar — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by denisbodnar — DISMISSED** [review]

File: server/routes/index.js [hunk 14]
```diff
@@ -12,7 +12,7 @@ module.exports = app => {
   app.get('/api/profile/:id', profileController.retrieve);
   app.put('/api/profile/:id', profileController.update);
 
-  app.post('/api/users', userController.create);
+  app.post('/api/auth/users', userController.create);
```
Comment (4asof4ik) [inline server/routes/index.js]: drop change

File: server/controllers/user.controller.js [hunk 15]
```diff
@@ -4,6 +4,12 @@ const User = require('../../config/db').User;
 const Profile = require('../../config/db').Profile;
 const Guest = require('../../config/db').Guest;
 const password = require('./../helper/passwordGenerator');
+const passwordHash = require('password-hash');
+const jwt = require('jsonwebtoken');
+const secret = require('./../../config/jwt.secretkey.json');
+const { mailer } = require('./../helper');
```
Comment (4asof4ik) [inline server/controllers/user.controller.js]: add messages and const... here

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

**Review by yarvanya — APPROVED** [review]

**Review by denisbodnar — APPROVED** [review]

**Review by 4asof4ik — APPROVED** [review]

**Review by yuliannakorol — APPROVED** [review]

**Pull request merged**